src/main/java/org/distributed/systems/chord/actors/Node.java

```diff
@@ -228,144 +276,119 @@ public Receive createReceive() {
+                System.out.println(address.getHostAddress());
+            } catch (Exception e) {
+                // TODO: need to handle
             }
         }
-        System.out.println("I accepted a Join Request from " + msg.requestorKey);
-        System.out.println("New Successor:" + this.sucessor.toString() + " with id:" + this.sucessorId);
-        System.out.println("New Predecessor:" + this.predecessor.toString() + " with id:" + this.predecessorId);
+
+        return "akka://ChordNetwork@" + centralEntityAddress + ":" + centralEntityAddressPort + "/user/ChordActor";
     }
 
+    private ActorRef closest_preceding_node(int id) {
+        for (int i = Node.m - 1; i >= 0; i--) {
+            if (this.fingerTable[i] == null)
+                continue;
+            if (isBetweenExeclusive(this.id, id, this.fingerTable[i].id))
+                return this.fingerTable[i].chordRef;
+        }
+        return getSelf();
+    }
 
-    /**
-     * Handles the case, when a requestor can be added as the sucessor of this node.
-     * This might be the case when:
-     * It's in between the node's id and the successor
-     * or when it's the requestor's key is the biggest id in the chord network.
-     * @param msg
-     * @throws Exception
-     */
-    private void handleJoinInsertAsSuccessor(JoinMessage.JoinRequest msg) throws Exception {
-        JoinMessage.JoinConfirmationRequest joinConfirmationRequestMessage = new JoinMessage.JoinConfirmationRequest(msg.requestor, msg.requestorKey, null, 0);
-        Timeout timeout = Timeout.create(Duration.ofMillis(ChordStart.STANDARD_TIME_OUT));
-        Future<Object> confirmationReqFuture = Patterns.ask(this.sucessor, joinConfirmationRequestMessage, timeout);
-        JoinMessage.JoinConfirmationReply result = (JoinMessage.JoinConfirmationReply) Await.result(confirmationReqFuture, timeout.duration());
-        //TODO: Handle timeout!
-        if (result.accepted) {
-            System.out.println("I return a successor" + this.sucessorId);
-            System.out.println("I am " + this.id);
-            JoinMessage.JoinReply joinReplyMessage = new JoinMessage.JoinReply(getSelf(), this.sucessor, true, this.id, this.sucessorId);
-            msg.requestor.tell(joinReplyMessage, getSelf());
-            this.sucessor = msg.requestor;
-            this.sucessorId = msg.requestorKey;
-            System.out.println("I confirmed the final join!");
-            System.out.println("My new successor: " + msg.requestorKey);
-            return;
-        } else {
-            JoinMessage.JoinReply joinReplyMessage = new JoinMessage.JoinReply(null, null, false);
-            msg.requestor.tell(joinReplyMessage, getSelf());
-            System.out.println("I declined the JOIN, predecessor rejected join!");
-            return;
+    private void fix_fingers() {
+        fix_fingers_next++;
+
+        if (fix_fingers_next > Node.m) {
+            fix_fingers_next = 1;
```

Comment:
This mean you will never update the first entry, a.k.a. your successor, with this method right?

Reply:
I think so, yes

Reply:
Well, let's think about this. Sucessor is updated on stableize. Maybe do a task for that one?